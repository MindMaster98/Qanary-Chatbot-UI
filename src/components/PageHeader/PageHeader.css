#header {
    position: relative;
    height: 75px;
    width: 100%;
    z-index: 1;
}

.questionsToggle {
    position: relative;
    min-width: 170px;
    height: 100%;
}

.questionsToggle:hover {
    background-color: #861f32;
    cursor: pointer;
}

.questionsToggle:focus {
    outline: none;
    background-color: #861f32;
    text-decoration: underline 1px #000 solid;
}

.exampleQuestion {
    width: 100%;
    display: flex;
    flex-wrap: wrap;
    justify-content: space-evenly;
    word-break: normal;
}