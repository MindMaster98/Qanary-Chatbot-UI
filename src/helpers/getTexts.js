let currentLanguage = window._env_?.DEFAULT_LANGUAGE;
let texts = require(`../texts/${currentLanguage}/texts.json`);
let listeners = [];

// This is not a typo, this is a generator function
// For more information, see here: https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Statements/function*
function* getIdGenerator() {
  let currentId = 0;
  while (true) {
    yield currentId;
    currentId++;
  }
}

const idGenerator = getIdGenerator();

function changeLanguage(newLanguage = "de") {
  if (currentLanguage === newLanguage) return;
  currentLanguage = newLanguage;
  try {
    texts = require(`../texts/${currentLanguage}/texts.json`);
    listeners.forEach((listener) => listener.callback());
  } catch (errorMessage) {
    console.error(errorMessage);
    currentLanguage = window._env_?.DEFAULT_LANGUAGE;
  }
}

function getCurrentLanguage() {
  return currentLanguage;
}

function getTexts() {
  return texts;
}

function addListener(callback) {
  const id = idGenerator.next().value;
  listeners.push({
    id,
    callback,
  });
  return id;
}

function removeListener(id) {
  const listenerPosition = listeners.indexOf((listener) => listener.id === id);
  listeners.splice(listenerPosition, 1);
}

export {
  changeLanguage,
  getTexts,
  addListener,
  removeListener,
  getCurrentLanguage,
};
