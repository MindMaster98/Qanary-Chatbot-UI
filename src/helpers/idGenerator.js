// This is not a typo, this is a generator function
// For more information, see here: https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Statements/function*
function* getIdGenerator() {
  let currentId = 0;
  while (true) {
    yield currentId;
    currentId++;
  }
}

export default getIdGenerator;
