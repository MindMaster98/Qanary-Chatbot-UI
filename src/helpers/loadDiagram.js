const d3 = window.d3;

export default function loadDiagram(data, svgId) {
  console.log(data, svgId);
  const dataPoints = data.data_points;
  // Convert data points from strings to objects
  const transformedDataPoints = dataPoints.map((dataPoint) => {
    const matches = dataPoint.match(/\(([0-9-.]*),(\d*)\)/);
    return {
      x: matches[1],
      y: matches[2],
    };
  });

  const margin = {
    left: 50,
    right: 25 + data["x-axis"].Label.length * 5,
    bottom: 30,
    top: 40,
  };

  // Get svg, its height and width
  const svg = d3.select(`#${svgId}`),
    width = svg.attr("width"),
    height = svg.attr("height");

  // Add X axis
  const xAxis = d3
    .scaleLinear()
    .domain([
      d3.min(transformedDataPoints, function (d) {
        return +d.x;
      }),
      d3.max(transformedDataPoints, function (d) {
        return +d.x;
      }),
    ])
    .range([margin.left, width - margin.right]);

  // Add Y axis
  const yAxis = d3
    .scaleLinear()
    .domain([
      d3.min(transformedDataPoints, function (d) {
        return +d.y;
      }),
      d3.max(transformedDataPoints, function (d) {
        return +d.y;
      }),
    ])
    .range([height - margin.bottom, margin.top]);

  // Set axes
  svg
    .append("g")
    .attr("transform", `translate(0,${height - margin.bottom})`)
    .call(d3.axisBottom(xAxis))
    .call((g) =>
      g
        .select(".tick:last-of-type text")
        .clone()
        .attr("x", (data["x-axis"].Label.length * 5))
        .attr("text-anchor", "start")
        .attr("font-weight", "bold")
        .text(data["x-axis"].Label)
    );
  svg
    .append("g")
    .attr("transform", `translate(${margin.left},0)`)
    .call(d3.axisLeft(yAxis))
    .call((g) =>
      g
        .select(".tick:last-of-type text")
        .clone()
        .attr("y", -30)
        .attr("text-anchor", "start")
        .attr("font-weight", "bold")
        .text(data["y-axis"].Label)
    );

  // Add the line
  svg
    .append("path")
    .datum(transformedDataPoints)
    .attr("fill", "none")
    .attr("stroke", "steelblue")
    .attr("stroke-width", 1.5)
    .attr(
      "d",
      d3
        .line()
        .x(function (d) {
          return xAxis(d.x);
        })
        .y(function (d) {
          return yAxis(d.y);
        })
    );
}
