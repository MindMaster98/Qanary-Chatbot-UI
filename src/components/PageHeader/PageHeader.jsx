import React, { Component } from "react";
import { Button, ClickableIcon, ContentContainer } from "..";
import { textsHelper } from "../../helpers";
import flag_icon from "../share/imgs/flag_icon.webp";
import info_icon_white from "../share/imgs/info_icon_white.svg";
import "./PageHeader.css";
import config from "../../config.json";

/**
 * @description This is the header of the page containing the title and various buttons
 *
 * @property {function} sendMessage The same function that's used in the MessageInput to send examplary questions
 */
export default class PageHeader extends Component {
  texts = textsHelper.getTexts();

  componentDidMount() {
    this.listenerId = textsHelper.addListener(() => {
      this.texts = textsHelper.getTexts();
      this.forceUpdate();
    });
  }

  componentWillUnmount() {
    textsHelper.removeListener(this.listenerId);
  }

  render() {
    return (
      <>
        <div tabIndex="-1" id="header">
          <ClickableIcon
            onClick={() => {
              const imprint = document.getElementById("imprint");
              const exemplaryQuestions = document.getElementById(
                "exemplaryQuestions"
              );
              if (!exemplaryQuestions.classList.contains("hidden")) {
                exemplaryQuestions.classList.toggle("hidden");
              }
              imprint.classList.toggle("hidden");
              imprint.focus();
            }}
            icon={info_icon_white}
            alt={this.texts["page-header"].icons["info-alt"]}
            style={{
              position: "relative",
              maxWidth: "100px",
              maxHeight: "40px",
              top: "calc((100% - 40px) / 2)",
            }}
          />
          <div className="pageTitle" tabIndex="0">
            <a
              href={config["chatbot-frontend-url"]}
              rel="noreferrer"
              target="_self"
              className="center white-text"
            >
              {this.texts["page-header"].title}
            </a>
          </div>
          <div
            className="questionsToggle"
            tabIndex="-1"
            onClick={() => {
              const exemplaryQuestion = document.getElementById(
                "exemplaryQuestions"
              );
              const imprint = document.getElementById("imprint");
              if (!imprint.classList.contains("hidden")) {
                imprint.classList.toggle("hidden");
              }
              exemplaryQuestion.classList.toggle("hidden");
              exemplaryQuestion.focus();
            }}
          >
            <span className="center white-text">
              {this.texts["options-menu"]["exemplary-questions"]}
            </span>
          </div>
          <ClickableIcon
            onClick={() => {
              if (textsHelper.getCurrentLanguage() === "de") {
                textsHelper.changeLanguage("en");
              } else {
                textsHelper.changeLanguage("de");
              }
            }}
            alt={this.texts["page-header"].icons["change-language-alt"]}
            icon={flag_icon}
            style={{
              position: "relative",
              maxWidth: "100px",
              maxHeight: "40px",
              top: "calc((100% - 40px) / 2)",
            }}
          />
        </div>
        <ContentContainer
          id="imprint"
          dangerouslySetInnerHTML={this.texts.credits}
        />
        <ContentContainer id="exemplaryQuestions">
          {this.texts["exemplary-questions"].map((questionText, i) => (
            // i is the shortest unique identifier in this case and the content will not be updated
            <div key={i} className="exampleQuestion">
              <div>{questionText}</div>
              <Button
                onClick={() => {
                  this.props.sendMessage(questionText);
                  document
                    .getElementById("exemplaryQuestions")
                    .classList.toggle("hidden");
                }}
                text={this.texts["example-questions"]["ask-question"]}
              />
            </div>
          ))}
        </ContentContainer>
      </>
    );
  }
}
