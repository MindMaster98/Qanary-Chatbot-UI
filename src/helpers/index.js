import * as textsHelper from './getTexts'

export { textsHelper }