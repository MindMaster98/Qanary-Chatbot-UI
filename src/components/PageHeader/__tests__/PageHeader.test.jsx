import renderer from "react-test-renderer";
import { act } from "@testing-library/react";
import ReactDOM from "react-dom";
import PageHeader from "../PageHeader";
import { textsHelper } from "../../../helpers"

// Snapshot tests

it("renders correctly with no props", () => {
  const tree = renderer.create(<PageHeader />).toJSON();
  expect(tree).toMatchSnapshot();
});

// Blackbox tests
// You cannot dispatch events using test renderer, so I need to actually add the element to the DOM
// and dispatch an actual event. Calling the function directly would oppose the idea of a blackbox.
// Why not enzyme? Because enzyme doesn't officially support React 17.

// This component isn't as obvious when it comes to testing. There's a lot that could happen and you could write
// a lot of tests. However, only few things are truly original functionality of this component. 
// The language toggle isn't original. It merely utilizes the getTexts module.
// Sending exemplary questions *is* original, but you need to mock the getTexts module to supply test questions first.
// Toggling the ContentContainer components *is* original.

let container;

beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

it("Imprint ContentContainer is toggled on icon click", () => {
  act(() => {
    ReactDOM.render(<PageHeader />, container);
  });

  const [ imprintIcon, ] = document.querySelectorAll(".clickableIcon");
  const imprintContainer = document.querySelector("#imprint");
  
  expect(imprintContainer.classList.contains("hidden")).toBeTruthy();

  imprintIcon.dispatchEvent(new InputEvent("click", { bubbles: true }));

  expect(imprintContainer.classList.contains("hidden")).toBeFalsy();

  imprintIcon.dispatchEvent(new InputEvent("click", { bubbles: true }));

  expect(imprintContainer.classList.contains("hidden")).toBeTruthy();
});

it("Imprint ContentContainer is not toggled on incorrect click", () => {
  act(() => {
    ReactDOM.render(<PageHeader />, container);
  });

  const questionsIcon = document.querySelector(".questionsToggle");
  const imprintContainer = document.querySelector("#imprint");
  
  expect(imprintContainer.classList.contains("hidden")).toBeTruthy();

  questionsIcon.dispatchEvent(new InputEvent("click", { bubbles: true }));

  expect(imprintContainer.classList.contains("hidden")).toBeTruthy();
});

it("Exemplary questions ContentContainer is toggled on div click", () => {
  act(() => {
    ReactDOM.render(<PageHeader />, container);
  });

  const questionsIcon = document.querySelector(".questionsToggle");
  const questionsContainer = document.querySelector("#exemplaryQuestions");
  
  expect(questionsContainer.classList.contains("hidden")).toBeTruthy();

  questionsIcon.dispatchEvent(new InputEvent("click", { bubbles: true }));

  expect(questionsContainer.classList.contains("hidden")).toBeFalsy();

  questionsIcon.dispatchEvent(new InputEvent("click", { bubbles: true }));

  expect(questionsContainer.classList.contains("hidden")).toBeTruthy();
});

it("Exemplary questions ContentContainer is not toggled on incorrect click", () => {
  act(() => {
    ReactDOM.render(<PageHeader />, container);
  });

  const [ imprintIcon, ] = document.querySelectorAll(".clickableIcon");
  const questionsContainer = document.querySelector("#exemplaryQuestions");
  
  expect(questionsContainer.classList.contains("hidden")).toBeTruthy();

  imprintIcon.dispatchEvent(new InputEvent("click", { bubbles: true }));

  expect(questionsContainer.classList.contains("hidden")).toBeTruthy();
});

it("You can send the exemplary questions as messages", () => {
  // This loads the mock module I created in ../../../helpers/__mocks__/getTexts.js
  jest.mock("../../../helpers/getTexts");
  const mockFn = jest.fn();

  act(() => {
    ReactDOM.render(<PageHeader sendMessage={mockFn} />, container);
  });

  const questionsContainer = document.querySelector("#exemplaryQuestions");
  const sendButton = questionsContainer.querySelector(".exampleQuestion > button");
  const exemplaryQuestion = textsHelper.getTexts()?.["exemplary-questions"]?.[0];
  questionsContainer.classList.toggle("hidden");
  
  sendButton.dispatchEvent(new InputEvent("click", { bubbles: true }));

  expect(mockFn).toHaveBeenCalledWith(exemplaryQuestion);
  // It's VERY important to unmock the module because the other tests rely on it being there
  jest.unmock("../../../helpers/getTexts");
})