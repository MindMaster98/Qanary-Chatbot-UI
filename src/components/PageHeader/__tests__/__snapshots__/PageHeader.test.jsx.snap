// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with no props 1`] = `
Array [
  <div
    id="header"
    tabIndex="-1"
  >
    <div
      className="clickableIcon "
      onClick={[Function]}
      style={
        Object {
          "maxHeight": "40px",
          "maxWidth": "100px",
          "position": "relative",
          "top": "calc((100% - 40px) / 2)",
        }
      }
      tabIndex="0"
    >
      <img
        alt="Informationen"
        src="info_icon_white.svg"
        tabIndex="-1"
      />
    </div>
    <div
      className="pageTitle"
      tabIndex="0"
    >
      <a
        className="center white-text"
        href="http://coronabot.ins.hs-anhalt.de/"
        rel="noreferrer"
        target="_self"
      >
        Coronabot Demonstrator
      </a>
    </div>
    <div
      className="questionsToggle"
      onClick={[Function]}
      tabIndex="-1"
    >
      <span
        className="center white-text"
      >
        Beispiele
      </span>
    </div>
    <div
      className="clickableIcon "
      onClick={[Function]}
      style={
        Object {
          "maxHeight": "40px",
          "maxWidth": "100px",
          "position": "relative",
          "top": "calc((100% - 40px) / 2)",
        }
      }
      tabIndex="0"
    >
      <img
        alt="Sprache wechseln"
        src="flag_icon.webp"
        tabIndex="-1"
      />
    </div>
  </div>,
  <div
    className="contentContainer hidden"
    dangerouslySetInnerHTML={
      Object {
        "__html": "Diese Anwendung wurde am <a href=\\"https://www.hs-anhalt.de/hochschule-anhalt/fachbereich-5/ueberblick.html\\" target=\\"_blank\\" rel=\\"noreferrer\\">Fachbereich Informatik und Sprachen</a> der <a href=\\"https://hs-anhalt.de/\\" target=\\"_blank\\" rel=\\"noreferrer\\">Hochschule Anhalt</a> im Rahmen der Lehrveranstaltung &quot;Question Answering &amp; Chatbots&quot; unter der Leitung von <a href=\\"https://www.hs-anhalt.de/hochschule-anhalt/service/personenverzeichnis/prof-dr-andreas-both.html\\" target=\\"_blank\\" rel=\\"noreferrer\\">Prof. Dr. Andreas Both</a> entwickelt. Die Betreuung erfolgte durch Prof. Both und <a href=\\"https://www.hs-anhalt.de/hochschule-anhalt/service/personenverzeichnis/aleksandr-perevalov.html\\" target=\\"_blank\\" rel\\"noreferrer\\">Aleksandr Perevalov</a> sowie durch die Projektpartner des <a href=\\"https://www.bmi.bund.de/\\" target=\\"_blank\\" rel=\\"noreferrer\\">Bundesministerium des Innern, für Bau und Heimat</a> und <a href=\\"https://www.itzbund.de/\\" target=\\"_blank\\" rel=\\"noreferrer\\">ITZBund</a>. Es handelt sich hierbei um einen wissenschaftlichen Demonstrator, der offizielle deutsche COVID-19 Chatbot ist <a href=\\"https://chatbot.it.bund.de/\\" target=\\"_blank\\" rel=\\"noreferrer\\">hier</a> abrufbar.<p><h3>Projektteam</h3>Das studentische Projektteam bestand aus Studierenden unserer Studieng&auml;nge &quot;Angewandte Informatik&quot; (Bachelor) und &quot;Data Science&quot; (Master): <ul><li>Backend: Johannes Richard Bartsch, Paul Heinze, Rostislav Iudin, Johannes Rudolf Herkner and Tim Schrader<li>Frontend: Jonas Wunsch</ul>",
      }
    }
    id="imprint"
    onBlur={[Function]}
    tabIndex="0"
  />,
  <div
    className="contentContainer hidden"
    id="exemplaryQuestions"
    onBlur={[Function]}
    tabIndex="0"
  >
    <div
      className="exampleQuestion"
    >
      <div>
        Tote Köln
      </div>
      <button
        className="button "
        onClick={[Function]}
        style={Object {}}
      >
        Diese Frage stellen
      </button>
    </div>
    <div
      className="exampleQuestion"
    >
      <div>
        Infektionszahlen Frankfurt
      </div>
      <button
        className="button "
        onClick={[Function]}
        style={Object {}}
      >
        Diese Frage stellen
      </button>
    </div>
    <div
      className="exampleQuestion"
    >
      <div>
        Wie viele Fälle gab es im Februar in Berlin Spandau?
      </div>
      <button
        className="button "
        onClick={[Function]}
        style={Object {}}
      >
        Diese Frage stellen
      </button>
    </div>
    <div
      className="exampleQuestion"
    >
      <div>
        Fälle 12.02.2021
      </div>
      <button
        className="button "
        onClick={[Function]}
        style={Object {}}
      >
        Diese Frage stellen
      </button>
    </div>
    <div
      className="exampleQuestion"
    >
      <div>
        Wie viele Fälle wurden diese Woche gemeldet?
      </div>
      <button
        className="button "
        onClick={[Function]}
        style={Object {}}
      >
        Diese Frage stellen
      </button>
    </div>
    <div
      className="exampleQuestion"
    >
      <div>
        Infektionszahlen Januar
      </div>
      <button
        className="button "
        onClick={[Function]}
        style={Object {}}
      >
        Diese Frage stellen
      </button>
    </div>
    <div
      className="exampleQuestion"
    >
      <div>
        Todeszahlen Oktober
      </div>
      <button
        className="button "
        onClick={[Function]}
        style={Object {}}
      >
        Diese Frage stellen
      </button>
    </div>
    <div
      className="exampleQuestion"
    >
      <div>
        Wie viele Tote gab es im Februar in Quedlinburg?
      </div>
      <button
        className="button "
        onClick={[Function]}
        style={Object {}}
      >
        Diese Frage stellen
      </button>
    </div>
    <div
      className="exampleQuestion"
    >
      <div>
        Fälle gestern
      </div>
      <button
        className="button "
        onClick={[Function]}
        style={Object {}}
      >
        Diese Frage stellen
      </button>
    </div>
    <div
      className="exampleQuestion"
    >
      <div>
        Wie viele Fälle wurden im Juni in Berlin und Brandenburg gemeldet?
      </div>
      <button
        className="button "
        onClick={[Function]}
        style={Object {}}
      >
        Diese Frage stellen
      </button>
    </div>
  </div>,
]
`;
